File: 0069. Sqrt(x)/main0069.java
public class main0069 {
    
}


class Solution {
    public int mySqrt(int x) {
        if (x < 2) {
            return x;
        }
        int lo = 2, hi = x / 2, mid = 2;
        long num;
        while (lo <= hi) {
            mid = lo + (hi - lo) / 2;
            num = (long) mid * mid;
            if (num == x) {
                return mid;
            } else if (num > x) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return hi;
    }
}

File: 0069. Sqrt(x)/main0069.cpp
#include <iostream>
#include <string>
#include<vector>
using namespace std;


class main0069 {
public:
  int mySqrt(int x) {      
    long l = 1;
    long r = static_cast<long>(x) + 1; // !!!! overflow
    while (l < r) {
      long m = l + (r - l) / 2;
      if (m * m > x) { 
        r = m;
      } else {
        l = m + 1;
      }
    }
    // l: smallest number such that l * l > x
    return l - 1;
  }
};

int main() {
    main0069 sol;
    int x = 4;
    int res = sol.mySqrt(x);
    cout << res;
}